java/arcs/core/host/ParticleContext.kt

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.host
+
+import arcs.core.data.Plan
+import arcs.core.entity.Handle
+import arcs.core.host.api.Particle
+import arcs.core.storage.StorageProxy.StorageEvent
+import arcs.core.util.TaggedLog
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+/** Maximum number of times a particle may fail to be started before giving up. */
+const val MAX_CONSECUTIVE_FAILURES = 5
+
+/**
+ * Holds per-[Particle] context state needed by [ArcHost] to implement [Particle] lifecycle.
```

Comment:
This is actually a bug we need to fix. It's possible for stopArc() to be called while startArc() is in progress. We might need to introduce more intermediate states to prevent this, like ArcState.STARTING or ArcState.STOPPING so that any ArcHost interface calls won't be re-entrant while another operation is in progress, which also suggests that we may need to queue them on the dispatcher or in a Flow.

Not for this PR, but worth noting that ArcHost.start/stop are not re-entrant safe for the same ArcId.